src/main/scala/com/microsoft/hyperspace/Hyperspace.scala

```diff
@@ -165,19 +171,31 @@ class Hyperspace(spark: SparkSession) {
   }
 }
 
-object Hyperspace {
-  private lazy val context = new ThreadLocal[HyperspaceContext]
+object Hyperspace extends Logging {
+  // Maintain HyperspaceContext for each Spark session, since Hyperspace depends on
+  // the session's properties such as configs, etc.
+  // Value as (thread id, hyperspace context) to enforce one client thread per Spark session
+  // as Hyperspace doesn't guarantee thread safety.
+  private[hyperspace] lazy val contexts =
+    new ConcurrentHashMap[SparkSession, (Long, HyperspaceContext)]().asScala
 
   private[hyperspace] def getContext(spark: SparkSession): HyperspaceContext = {
-    val ctx = context.get()
-    if (ctx == null || !ctx.spark.equals(spark)) {
-      // Ensure the Spark session the caller provides is the same as
-      // the one HyperspaceContext is using because Hyperspace depends on the
-      // session's properties such as configs, etc.
-      context.set(new HyperspaceContext(spark))
+    val ctx = contexts.get(spark)
+    if (!ctx.isDefined) {
+      contexts.synchronized {
+        if (contexts.size >= HyperspaceConf.maxSupportedSparkSessions(spark)) {
+          throw HyperspaceException("Too many Spark sessions uses Hyperspace." +
+            "Please increase the config: spark.hyperspace.maxSessions")
+        }
+        contexts.put(spark, (Thread.currentThread().getId, new HyperspaceContext(spark)))
+      }
+    } else if (ctx.get._1 != Thread.currentThread().getId) {
+      logWarning(s"Hyperspace is not thread safe for threads using one Spark session. " +
+        "Please be aware of it. Current thread id: ${Thread.currentThread().getId}, " +
```

Comment:
What is the implication of potentially sharing the `HyperspaceContext` by threads if we already know that it's not thread safe? Isn't it better to fail faster here instead of failing somewhere else?

Reply:
checkAnswer won't work if we throw an exception here. No idea how to fix it and one possible scenario:
- setup & check in main thread
- run with a worker thread

Reply:
are you still blocked on this?

Reply:
gentle ping

Reply:
I think throwing exception is too restrictive. Users might not know how to fix the issue ..?

Reply:
How about this?

> Isn't it better to fail faster here instead of failing somewhere else?

You can't suggest this is not supported? Being restrictive seems better than failing at a random place because Hyperspace is not thread safe?

Reply:
I thought it's user's responsibility to use the API which is not thread safe. 
Different thread id not always means they are running concurrently.. 
I think we need documentation rather than restricting the use case?

Reply:
and I got the following error after throwing an exception here ..  ?
```
[info] - Verify createIndex and refreshIndex on Delta Lake table. *** FAILED ***
[info]   com.microsoft.hyperspace.HyperspaceException: Only creating index over HDFS file based scan nodes is supported.
[info]   at com.microsoft.hyperspace.actions.CreateAction.validate(CreateAction.scala:49)
[info]   at com.microsoft.hyperspace.actions.Action.run(Action.scala:88)
[info]   at com.microsoft.hyperspace.actions.Action.run$(Action.scala:83)
[info]   at com.microsoft.hyperspace.actions.CreateAction.run(CreateAction.scala:30)
[info]   at com.microsoft.hyperspace.index.IndexCollectionManager.create(IndexCollectionManager.scala:47)
[info]   at com.microsoft.hyperspace.index.CachingIndexCollectionManager.create(CachingIndexCollectionManager.scala:78)
[info]   at com.microsoft.hyperspace.Hyperspace.createIndex(Hyperspace.scala:49)
[info]   at com.microsoft.hyperspace.index.DeltaLakeIntegrationTest.$anonfun$new$4(DeltaLakeIntegrationTest.scala:66)
[info]   at com.microsoft.hyperspace.index.DeltaLakeIntegrationTest.$anonfun$new$4$adapted(DeltaLakeIntegrationTest.scala:60)
[info]   at com.microsoft.hyperspace.index.HyperspaceSuite.withTempPathAsString(HyperspaceSuite.scala:120)
```

Reply:
> Different thread id not always means they are running concurrently..

But the new structure allows multiple threads accessing the object concurrently? The existing implementation makes sure only one to one mapping via thread local. So, I would keep this condition as it is, and if there are demands for accessing the hyperspace object from multiple threads, we should think about making it thread-safe instead of documenting.

Plus, I think random failures are much worse and hard to debug, esp. failures caused by the thread safety.

Reply:
I agree on how difficult analyzing failures from concurrent threads. 
I added the exception change, then this can be a breaking change for some use case.

Reply:
Added breaking change label so we can document it in the release notes.

Reply:
btw, even if this is a breaking change, it's alerting the user a possible misuse where hyperspace context is being recreated, I think it's worthwhile.